### PR TITLE
Two small fixes (re. FP32-output-logit&residuals) from Nemotron

### PR DESCRIPTION
**Pull request by kwangjunahn**
- [x] I, the PR author, have personally reviewed every line of this PR.

## This PR = two small fixes from Nemotron

This PR adapts downstream fixes (from nv-mistral-megatron) authored by Roger Waleffe to current upstream Megatron-LM.

| Downstream source | Commits in this PR | What's ported in this PR |
| --- | --- | --- |
| [`b6a5640`](https://github.com/NVIDIA-NeMo/nv-mistralai-megatron/commit/b6a5640432ba345a3073c84ef731a21554209cb5)  | [`9f872c98`](https://github.com/kwangjunahn/Megatron-LM/commit/9f872c986ec478e18dcf73d12b51a33f8c919e5e), [`a1ba05b6`](https://github.com/kwangjunahn/Megatron-LM/commit/a1ba05b694a852ca061794f0077f1eab4370e2cf), [`4907ed9c`](https://github.com/kwangjunahn/Megatron-LM/commit/4907ed9ca71540338f279bc023cb828cc2323512) | FP32 logit output ownership, per-source MIMO bridge dtypes, local embedding promotion, provider wiring, and adapted/extended tests. |
| [`092ff0a`](https://github.com/NVIDIA-NeMo/nv-mistralai-megatron/commit/092ff0acb1ffe32f0c3345b5a56f321c253d9592)<sup>*</sup> | [`a1ba05b6`](https://github.com/kwangjunahn/Megatron-LM/commit/a1ba05b694a852ca061794f0077f1eab4370e2cf) | Port only the precision reset in `examples/mimo/model_providers/radio_encoder.py::_make_dense_non_hybrid`. It keeps copied RADIO tower/projector configs in parameter dtype instead of inheriting the language model's FP32 residual policy. |

<sup>*</sup> No Wide Residual, ShortcutMoE, MTP, or inference behavior from `092ff0a` is included in this PR; those changes will be incorporated separately in existing Wide Residual PRs.

# What does this PR do?

Fixes two precision-boundary problems exposed when a language model uses FP32 residual accumulation:

1. Makes mixed-dtype tensor-parallel output logits safe for cross entropy's in-place workspace reuse.
2. Keeps MIMO modality towers and bridge communication in parameter dtype, promoting activations only when they enter the FP32 language embedding stream.

## FP32 output logits

Previously, the Transformer Engine GEMM returned a two-dimensional custom-autograd output that `_linear_forward` reshaped into the final logit tensor. The reshaped result was a view, which could not safely be modified in place by cross entropy.

The new path allocates the final-shaped output first, gives TE a two-dimensional view through its `out` argument, and returns the owning tensor. This does not introduce a clone.

## MIMO precision flow

`fp32_residual_connection` promotes the language pipeline dtype to FP32. Modality configs copied from that language config must not inherit this language-only policy: the pretrained modality tower and projector remain in parameter dtype.

`MultiModulePipelineCommunicator` (in `megatron/core/pipeline_parallel/multimodule_communicator.py`) now accepts an optional `bridge_comm_dtypes` mapping keyed by source module. The Nemotron MIMO provider configures the RADIO encoder bridge with:

```python
bridge_comm_dtypes={
    RADIO_ENCODER_MODULE_NAME: language_config.params_dtype,
}
```

(See `examples/mimo/model_providers/nemotron_moe_vlm.py`.)

With all of this, now the precision flow looks like:

```text
Forward

Modality tower/projector
         BF16
           |
           | BF16 bridge
           | sender dtype validated
           v
MIMO embedding merge
     BF16 -> FP32
           |
           v
Language embedding/residual stream
         FP32
```

```text
Backward

FP32 merge gradient
         |
         | autograd through the local cast
         v
BF16 bridge gradient
         |
         v
BF16 modality source
```

This avoids widening modality communication to FP32 while preserving the FP32 language residual stream.

## Summary of what's implemented in this PR

- Preallocate the final-shaped tensor-parallel linear output and return its owning tensor.
- Add optional per-source `bridge_comm_dtypes` to `MultiModulePipelineCommunicator`.
- Validate all bridge send paths against the receive dtype before entering NCCL. (See `megatron/core/pipeline_parallel/bridge_communicator.py`.)
- Cast modality embeddings to the destination language embedding dtype at the local MIMO merge.
- Keep copied RADIO tower/projector configs in parameter dtype.
- Propagate `--logit-dtype` into the Nemotron MoE VLM language model spec.
- Configure the RADIO bridge from the source parameter dtype.
 

## Issue tracking

Linked issue: N/A - bug-fix port of the downstream commits listed above.

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] GPU-kernel determinism testing is not applicable; this PR does not change a kernel or dispatch policy
- [x] I have added proper typing to my code
- [x] I have documented the new optional communicator argument and precision contract
- [x] I have run `tools/autoformat.sh` on the final PR head

### CODE REVIEW

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/models/mimo/model/base.py [hunk 1]
```diff
@@ -272,6 +272,9 @@ def align_embeddings_by_token_positions(
                 modality_embeddings, modality_token_indices, batch_size * seq_length
             )
             for modality_name, modality_emb in modality_embeddings.items():
+                # Bridges keep modality activations in parameter dtype; promote only when
+                # they join a higher-precision language-model residual stream.
+                modality_emb = modality_emb.to(dtype=dtype)
```
Comment (yashaswikarnati) [inline megatron/core/models/mimo/model/base.py:277]: just for my own understanding, curious why we would need this dtype cast here

Comment (kwangjunahn) [reply megatron/core/models/mimo/model/base.py:277]: The combined embedding tensor uses the main LM embedding/residual-stream dtype, which is FP32 with FP32 residuals. We keep modality communication in BF16 and promote it only at the local merge boundary; backward then returns a BF16 gradient through the cast. Does this make sense?

**Review by Phlip79 — APPROVED** [review]

**Comment by jaredcasper** [comment]
/ok to test [4907ed9](https://github.com/NVIDIA/Megatron-LM/pull/7480/commits/4907ed9ca71540338f279bc023cb828cc2323512)

File: megatron/core/models/mimo/model/base.py (hunk 1 above)
Comment (yashaswikarnati) [reply megatron/core/models/mimo/model/base.py:277]: thanks for the info! lgtm

**Review by yashaswikarnati — APPROVED** [review]
